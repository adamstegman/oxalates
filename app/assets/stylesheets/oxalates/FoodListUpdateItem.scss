@import "./body-form";
@import "./error";

.edit-food {
  @extend %body-form;
}

.error {
  @extend %error;
}
