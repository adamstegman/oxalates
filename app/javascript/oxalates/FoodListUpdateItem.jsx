import React from 'react';
import PropTypes from 'prop-types';
import { foodPropType } from './foodPropType';

export const FoodListUpdateItem = ({
  editingFood,
  editingFoodListId,
  cancelEditingFood,
  setEditingFood,
  updateFood,
  error,
  password,
}) => {
  const setFoodName = event => {
    const foodWithName = Object.assign({}, editingFood, { name: event.target.value });
    setEditingFood(foodWithName, editingFoodListId);
  };
  const setFoodOxalateMg = event => {
    const foodWithOxalateMg = Object.assign({}, editingFood, { oxalateMg: event.target.value });
    setEditingFood(foodWithOxalateMg, editingFoodListId);
  };
  const setFoodServing = event => {
    const foodWithServing = Object.assign({}, editingFood, { serving: event.target.value });
    setEditingFood(foodWithServing, editingFoodListId);
  };
  const _updateFood = event => {
    event.preventDefault();
    updateFood(password, editingFood, editingFoodListId);
  };

  const {
    name,
    oxalateMg,
    serving,
  } = editingFood;
  const errorMessages = (error || []).map((msg, i) => <p key={i} className="error">{msg}</p>);
  return (
    <form className="edit-food" onSubmit={_updateFood}>
      <label htmlFor="name">Name</label>
      <input type="text" name="name" autoFocus value={name || ''} onChange={setFoodName} />
      <label htmlFor="oxalateMg">Oxalates per serving (mg)</label>
      <input type="number" name="oxalateMg" step="any" value={oxalateMg || 0.0} onChange={setFoodOxalateMg} />
      <label htmlFor="serving">Serving</label>
      <input type="text" name="serving" value={serving || ''} onChange={setFoodServing} />
      {errorMessages}
      <input type="submit" value="Update" />
      <button type="button" onClick={cancelEditingFood}>Cancel</button>
    </form>
  );
};

FoodListUpdateItem.propTypes = {
  editingFood: foodPropType.isRequired,
  editingFoodListId: PropTypes.node.isRequired,
  updateFood: PropTypes.func.isRequired,
  cancelEditingFood: PropTypes.func.isRequired,
  setEditingFood: PropTypes.func.isRequired,
  error: PropTypes.arrayOf(PropTypes.string),
  password: PropTypes.string,
};
