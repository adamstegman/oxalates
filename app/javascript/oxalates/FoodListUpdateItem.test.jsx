import React from 'react';
import renderer from 'react-test-renderer';

import { FoodListUpdateItem } from './FoodListUpdateItem';

const editingFood = {
  id: 1,
  name: 'some food',
  serving: 'some serving',
  oxalateMg: 3.14,
};

test('FoodListUpdateItem renders a pre-filled form', () => {
  const component = renderer.create(
    <FoodListUpdateItem editingFood={editingFood} editingFoodListId={1} updateFood={() => {}} setEditingFood={() => {}} />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('FoodListUpdateItem renders a failed form', () => {
  const component = renderer.create(
    <FoodListUpdateItem editingFood={editingFood} editingFoodListId={1} updateFood={() => {}} setEditingFood={() => {}} error={['error 1']} />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('FoodListUpdateItem renders a server error', () => {
  const component = renderer.create(
    <FoodListUpdateItem editingFood={editingFood} editingFoodListId={1} updateFood={() => {}} setEditingFood={() => {}} error={'error message'} />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
