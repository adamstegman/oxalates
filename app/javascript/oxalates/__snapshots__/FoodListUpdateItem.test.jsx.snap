// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FoodListUpdateItem renders a failed form 1`] = `
<form
  className="edit-food"
  onSubmit={[Function]}
>
  <label
    htmlFor="name"
  >
    Name
  </label>
  <input
    autoFocus={true}
    name="name"
    onChange={[Function]}
    type="text"
    value="some food"
  />
  <label
    htmlFor="oxalateMg"
  >
    Oxalates per serving (mg)
  </label>
  <input
    name="oxalateMg"
    onChange={[Function]}
    step="any"
    type="number"
    value={3.14}
  />
  <label
    htmlFor="serving"
  >
    Serving
  </label>
  <input
    name="serving"
    onChange={[Function]}
    type="text"
    value="some serving"
  />
  <p
    className="error"
  >
    error 1
  </p>
  <input
    type="submit"
    value="Update"
  />
</form>
`;

exports[`FoodListUpdateItem renders a pre-filled form 1`] = `
<form
  className="edit-food"
  onSubmit={[Function]}
>
  <label
    htmlFor="name"
  >
    Name
  </label>
  <input
    autoFocus={true}
    name="name"
    onChange={[Function]}
    type="text"
    value="some food"
  />
  <label
    htmlFor="oxalateMg"
  >
    Oxalates per serving (mg)
  </label>
  <input
    name="oxalateMg"
    onChange={[Function]}
    step="any"
    type="number"
    value={3.14}
  />
  <label
    htmlFor="serving"
  >
    Serving
  </label>
  <input
    name="serving"
    onChange={[Function]}
    type="text"
    value="some serving"
  />
  <input
    type="submit"
    value="Update"
  />
</form>
`;

exports[`FoodListUpdateItem renders a server error 1`] = `
<form
  className="edit-food"
  onSubmit={[Function]}
>
  <label
    htmlFor="name"
  >
    Name
  </label>
  <input
    autoFocus={true}
    name="name"
    onChange={[Function]}
    type="text"
    value="some food"
  />
  <label
    htmlFor="oxalateMg"
  >
    Oxalates per serving (mg)
  </label>
  <input
    name="oxalateMg"
    onChange={[Function]}
    step="any"
    type="number"
    value={3.14}
  />
  <label
    htmlFor="serving"
  >
    Serving
  </label>
  <input
    name="serving"
    onChange={[Function]}
    type="text"
    value="some serving"
  />
  <p
    className="error"
  >
    error message
  </p>
  <input
    type="submit"
    value="Update"
  />
</form>
`;
